### PR TITLE
Improve span handling performance (setTag and serialization)

### PR DESCRIPTION
**Pull request by pawelchcki**
### Description

Use direct value access instead of getters and setters to improve the performance of span handling.

Additionally:
- ~cache EnvVariableRegistry variable conversion result~
- improve p50 setTag performance by using associative array to filter out edge cases. 

The performance measuring env graphs seem to be happy:
image

### Readiness checklist
- [x] [Changelog entry](docs/changelog.md) added, if necessary
- [x] Tests added for this feature/bug

### CODE REVIEW

File: src/DDTrace/Encoders/SpanEncoder.php [hunk 1]
```diff
@@ -15,43 +16,43 @@ final class SpanEncoder
      * @param Span $span
      * @return array
      */
-    public static function encode(Span $span)
+    public static function encode(SpanData $span)
```
Comment (SammyK) [inline src/DDTrace/Encoders/SpanEncoder.php]: ❤️

File: src/DDTrace/Encoders/SpanEncoder.php [hunk 2]
```diff
@@ -15,43 +16,43 @@ final class SpanEncoder
...
 
-        if ($span->getType() !== null) {
-            $arraySpan['type'] = $span->getType();
+        if ($span->type !== null) {
+            $arraySpan['type'] = $span->type;
         }
 
-        if ($span->isFinished()) {
-            $arraySpan['duration'] = (int) ($span->getDuration() . '000');
+        if ($span->duration != null) { // is span finished ?
```
Comment (SammyK) [inline src/DDTrace/Encoders/SpanEncoder.php]: I'd do a strict check here. :)

```suggestion
        if ($span->duration !== null) { // is span finished ?
```

File: src/DDTrace/Processing/TraceAnalyticsProcessor.php [hunk 3]
```diff
@@ -13,33 +13,48 @@
  */
 final class TraceAnalyticsProcessor
 {
+    /**
+    * @param bool|float $value
+    */
+    public static function normalizeAnalyticsValue(&$array, $value)
```
Comment (SammyK) [inline src/DDTrace/Processing/TraceAnalyticsProcessor.php]: I think I'd call this `$metrics` and update the DocBlock. :)

```suggestion
    public static function normalizeAnalyticsValue(&$metrics, $value)
```

File: src/DDTrace/Span.php [hunk 4]
```diff
@@ -4,139 +4,71 @@
 
 use DDTrace\Integrations\Integration;
 use DDTrace\Contracts\Span as SpanInterface;
```
Comment (SammyK) [inline src/DDTrace/Span.php]: We can remove this :)

File: src/DDTrace/Span.php [hunk 5]
```diff
@@ -4,139 +4,71 @@
...
+use DDTrace\Data\Span as SpanData;
+
 use DDTrace\Contracts\SpanContext as SpanContextInterface;
```
Comment (SammyK) [inline src/DDTrace/Span.php]: We can remove this line as well as the new line above it :)

File: src/DDTrace/Span.php [hunk 6]
```diff
@@ -4,139 +4,71 @@
...
 use DDTrace\Exceptions\InvalidSpanArgument;
+use DDTrace\SpanContext as SpanContext;
```
Comment (SammyK) [inline src/DDTrace/Span.php]: This is redundant so we can remove this as well :)

File: src/DDTrace/Span.php [hunk 7]
```diff
@@ -4,139 +4,71 @@
...
-    /**
-     * @var Integration
-     */
-    private $integration = null;
-
-    /**
-     * @var bool Whether or not this trace can be even considered for trace analytics automatic configuration.
-     */
-    private $isTraceAnalyticsCandidate = false;
+    private static $METRIC_NAMES = [ Tag::ANALYTICS_KEY => true ];
```
Comment (SammyK) [inline src/DDTrace/Span.php]: Just a nitpick. I know these kind of act more like constants, hence the name in ALL CAPS and snake_case, but I'd still have this follow how other class property names using camelCase. And I'd do the same for `$SPECIAL_TAGS` below. Again - that's just my opinion, you don't have to change anything. :)

File: src/DDTrace/Span.php [hunk 8]
```diff
@@ -192,48 +124,49 @@ public function getDuration()
...
-
         if ($key !== (string)$key) {
             throw InvalidSpanArgument::forTagKey($key);
         }
 
-        if ($key === Tag::ERROR) {
-            $this->setError($value);
-            return;
-        }
+        if (array_key_exists($key, self::$SPECIAL_TAGS)) {
```
Comment (SammyK) [inline src/DDTrace/Span.php]: 🔝

File: src/DDTrace/SpanContext.php [hunk 9]
```diff
@@ -174,6 +126,13 @@ public function getAllBaggageItems()
 
     public function isEqual(SpanContextInterface $spanContext)
     {
+        if (is_subclass_of($spanContext, "DDTrace\Data\SpanContext")) {
```
Comment (SammyK) [inline src/DDTrace/SpanContext.php]: We can remove a function invocation here. :)

```suggestion
        if ($spanContext instanceof SpanContextData) {
```

File: src/DDTrace/SpanContext.php [hunk 10]
```diff
@@ -174,6 +126,13 @@ public function getAllBaggageItems()
...
+            return $this->traceId == $spanContext->traceId
```
Comment (SammyK) [inline src/DDTrace/SpanContext.php]: Just because PHP's type coercion can get squirrely. :)

```suggestion
            return $this->traceId === $spanContext->traceId
```

File: src/DDTrace/Tracer.php [hunk 11]
```diff
@@ -19,7 +19,10 @@
 use DDTrace\Contracts\Scope as ScopeInterface;
 use DDTrace\Contracts\Span as SpanInterface;
 use DDTrace\Contracts\SpanContext as SpanContextInterface;
+use DDTrace\Data\Span as SpanData;
+use DDTrace\Data\SpanContext as SpanContextData;
```
Comment (SammyK) [inline src/DDTrace/Tracer.php]: We can remove these two lines and the `use DDTrace\Time;` line below. :)

File: src/DDTrace/Configuration/EnvVariableRegistry.php [hunk 12]
```diff
@@ -8,6 +8,11 @@
  */
 class EnvVariableRegistry implements Registry
 {
+    /**
+     * @var array
+     */
+    private $keyTranslator = [];
```
Comment (labbati) [inline src/DDTrace/Configuration/EnvVariableRegistry.php]: I am actually quite surprised that this brings a noticeable performance improvement. The reason for saying this is that this registry is a very low level one that is only called once per key (at least in theory). In fact, the upstream classes cache the results. Not saying we should not so it (actually some values are not cached on purpose) but still I am surprised :)

File: src/DDTrace/Data/Span.php [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Resource is a query to a service. A web application might use
+     * resources like "/user/{user_id}". A sql database might use resources
+     * like "select * from user where id = ?".
+     *
+     * You can track thousands of resources (not millions or billions) so
+     * prefer normalized resources like "/user/{id}" to "/user/123".
+     *
+     * Resources should only be set on an app's top level spans.
```
Comment (labbati) [inline src/DDTrace/Data/Span.php]: > Resources should only be set on an app's top level spans.

Not sure about this sentence, there are non-top level spans that are perfectly fit for a resource, e.q. `sql.query`. Wouldn't you agree?

File: src/DDTrace/Data/Span.php [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * @var string
+     */
+    public $resource;
+
+    /**
+     * Service is the name of the process doing a particular job. Some
+     * examples might be "user-database" or "datadog-web-app". Services
+     * will be inherited from parents, so only set this in your app's
+     * top level span.
```
Comment (labbati) [inline src/DDTrace/Data/Span.php]: Again I am not sure that this is the spirit of the service. Why should a PDO span inherit the service name `payments-api` instead of having the service `pdo`?

File: src/DDTrace/Data/SpanContext.php [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace DDTrace\Data;
+
+use DDTrace\Contracts\SpanContext as SpanContextInterface;
+
+abstract class SpanContext implements SpanContextInterface
+{
+    /**
+     * The unique integer (64-bit unsigned) ID of the trace containing this span.
+     * It is stored in hexadecimal representation.
```
Comment (labbati) [inline src/DDTrace/Data/SpanContext.php]: This is unsigned 63 bits: [link](https://github.com/DataDog/dd-trace-php/blob/1665e91a9ed9717f2c5906c9a11e27981620a862/src/DDTrace/ID.php#L12).

Also, I don't think that this is stored as hexadecimal. The [generate id](https://github.com/DataDog/dd-trace-php/blob/1665e91a9ed9717f2c5906c9a11e27981620a862/src/DDTrace/ID.php#L12) is the string representation of a 63-bit unsigned integer number `"1234567"` and it is [stored as it is](https://github.com/DataDog/dd-trace-php/blob/1665e91a9ed9717f2c5906c9a11e27981620a862/src/DDTrace/SpanContext.php#L37)

File: src/DDTrace/Data/SpanContext.php [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+     *
+     * @var string
+     */
+    public $traceId;
+
+    /**
+     * The span integer (64-bit unsigned) ID.
```
Comment (labbati) [inline src/DDTrace/Data/SpanContext.php]: Same as above

File: src/DDTrace/Data/SpanContext.php [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+     * It is stored in hexadecimal representation.
+     *
+     * @var string
+     */
+    public $spanId;
+
+    /**
+     * The span integer ID of the parent span.
+     * It is stored in hexadecimal representation.
```
Comment (labbati) [inline src/DDTrace/Data/SpanContext.php]: Same as above

File: src/DDTrace/Encoders/SpanEncoder.php [hunk 18]
```diff
@@ -15,43 +16,43 @@ final class SpanEncoder
      * @param Span $span
```
Comment (labbati) [inline src/DDTrace/Encoders/SpanEncoder.php]: This is `SpanData` now

File: src/DDTrace/Configuration/EnvVariableRegistry.php [hunk 19]
```diff
@@ -35,7 +40,15 @@ public function __construct($prefix = 'DD_')
...
-        $value = getenv($this->convertKeyToEnvVariableName($key));
+        if (array_key_exists($key, $this->keyTranslator)) {
+            $key = $this->keyTranslator[$key];
+        } else {
+            $new_key = $this->convertKeyToEnvVariableName($key);
+            $this->keyTranslator[$key] = $new_key;
+            $key = $new_key;
+        }
+
+        $value = getenv($key);
```
[the remark this answers is not in the document]
Comment (pawelchcki) [reply src/DDTrace/Configuration/EnvVariableRegistry.php]: Agreed its much cleaner . Seems my performance hat squeezed my head maybe a little too tight :)
And I wrote it like that to avoid extra array lookup :D.
However it does look like my version produces 5 more opcodes to process. I think going with your approach with one extra array lookup is less expensive than processing more opcodes. And cleaner PHP too :)

File: src/DDTrace/Processing/TraceAnalyticsProcessor.php [hunk 20]
```diff
@@ -13,33 +13,48 @@
...
+    {
+        if (true === $value || null === $value) {
```
Comment (labbati) [inline src/DDTrace/Processing/TraceAnalyticsProcessor.php]: Why should we convert `null` to `100%`? If the reason is, as how I understand it, to [remove this](https://github.com/DataDog/dd-trace-php/pull/398/files#diff-7952f110894c1eceb5170b6255a3550dL42) then I believe we should instead set the metric over there and remove this risky `null === $value` which looks a little bit counter intuitive to me. WDYT?

File: src/DDTrace/Configuration/EnvVariableRegistry.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (pawelchcki) [reply src/DDTrace/Configuration/EnvVariableRegistry.php]: Actually Just noticed this key translation is actually cached already :) by all the (bool|string|...l)Value thats actually used to access the values. 

Probably the only way to speed it up would be to use hashmap thats shared between requests.

File: src/DDTrace/Span.php (hunk 7 above)
Comment (pawelchcki) [reply src/DDTrace/Span.php]: Conventions are there to be broken 🙈 
Yeah it was a `const` but 5.4 made me change it to this :(.

I'll follow your suggestion though

File: src/DDTrace/Configuration/EnvVariableRegistry.php (hunk 12 above)
Comment (pawelchcki) [reply src/DDTrace/Configuration/EnvVariableRegistry.php]: Indeed. I've seen on the Profiler it being shown quite high. TBF I haven't checked this after changes if it improved the performance significantly. I've since now removed it when I noticed the boolValue et. al does indeed cache it. https://github.com/DataDog/dd-trace-php/pull/398#discussion_r274365118

File: src/DDTrace/Data/Span.php (hunk 13 above)
Comment (pawelchcki) [reply src/DDTrace/Data/Span.php]: Agreed!

File: src/DDTrace/Data/Span.php (hunk 14 above)
Comment (pawelchcki) [reply src/DDTrace/Data/Span.php]: 👍

File: src/DDTrace/Data/SpanContext.php (hunk 15 above)
Comment (pawelchcki) [reply src/DDTrace/Data/SpanContext.php]: 👍

File: src/DDTrace/Processing/TraceAnalyticsProcessor.php (hunk 20 above)
Comment (pawelchcki) [reply src/DDTrace/Processing/TraceAnalyticsProcessor.php]: I don't know ;) I only copy pasted it here :D 

Let me take a look :)

Comment (pawelchcki) [reply src/DDTrace/Processing/TraceAnalyticsProcessor.php]: I agree the `null === $value` looks weird, let me remvoe it and see what breaks.

**Comment by pawelchcki** [comment]
Thanks @labbati and @SammyK for the reviews!
lets :shipit:
